src/db.c

```diff
@@ -792,34 +792,53 @@ void keysCommand(client *c) {
     setDeferredArrayLen(c,replylen,numkeys);
 }
 
+/* Data used by the dict scan callback. */
+typedef struct {
+    list *keys;   /* elements that collect from dict*/
+    robj *o;      /* o must be a hash/set/zset object, NULL means current db*/
+    sds typename; /* the particular type when scan the db */
+    sds pattern;  /* pattern string, NULL means no pattern */
+    long sampled; /* cumulative number of keys sampled */
+} scanData;
+
 /* This callback is used by scanGenericCommand in order to collect elements
  * returned by the dictionary iterator into a list. */
 void scanCallback(void *privdata, const dictEntry *de) {
-    void **pd = (void**) privdata;
-    list *keys = pd[0];
-    robj *o = pd[1];
-    robj *key, *val = NULL;
+    scanData *data = (scanData *)privdata;
+    list *keys = data->keys;
+    robj *o = data->o;
+    robj *val = NULL;
+    data->sampled++;
+
+    /* typename is only meaningful when o is NULL*/
+    serverAssert(!data->typename || !o);
```

Comment:
Not sure the assert is right.
Now it is true if one of them is not NULL, but as the comment,
it should be `serverAssert(!o || data->typename)`

Reply:
If so, it should be `serverAssert(!(o && data->typename))`.

Reply:
I think `!data->typename || !o` and `!(o && data->typename)` are the same statement，does I miss something？But the `!(o && data->typename)` is easier to understand, I prefer to use your sentence.

Reply:
I think `!data->typename || !o` and `!(o && data->typename)` are the same statement，does I miss something？But the `!(o && data->typename)` is easier to understand, I prefer to use your sentence.

Reply:
Ohh, yes, they are the same.

Reply:
I have copied your code， thank you